src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/Validator.cs

```diff
@@ -518,7 +518,7 @@ private static void EnsureValidPropertyType(string propertyName, Type propertyTy
         private static ICollection<KeyValuePair<ValidationContext, object?>> GetPropertyValues(object instance,
             ValidationContext validationContext)
         {
-            var properties = TypeDescriptor.GetProperties(instance);
+            var properties = TypeDescriptor.GetProperties(instance.GetType());
```

Comment:
But this will break instance-registered type descriptors...

```cs
TypeDescriptor.Add(..., instance);
Validator.TryValidateObject(instance, ...);
```

Reply:
But how would this ever work, since everything in `ValidationAttributeStore` is based only on the type?

Reply:
For example, you can have a separate buddy type that you use for certain objects:

```cs
TypeDescriptor.AddProvider(
    new AssociatedMetadataTypeTypeDescriptionProvider(
        typeof(User),
        typeof(SpecialUserMetadata)),
    specialUser);
```
```cs
class SpecialUserMetadata
{
    // Special users can have really long usernames.
    [StringLength(1000)]
    public string Username { get; set; }
}
```

Reply:
@bricelam I'm not sure how to use this. Running on .NET Framework 4.8 with the code below, the output is:

```
True
0
```

That is, the validation doesn't happen. Changing:

```C#
TypeDescriptor.AddProvider(
    new AssociatedMetadataTypeTypeDescriptionProvider(
        typeof(ThisIsTheEntity),
        typeof(ThisIsTheBuddy)),
    specialUser);
```

to

```C#
TypeDescriptor.AddProvider(
    new AssociatedMetadataTypeTypeDescriptionProvider(
        typeof(ThisIsTheEntity),
        typeof(ThisIsTheBuddy)),
    new ThisIsTheEntity());
```

Does result in validation happening.

Can you provide more information on how passing the instance is supposed to work? Is there something wrong with my test code?

```C#
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;

namespace ConsoleApplication1
{
    internal class Program
    {
        public static void Main(string[] args)
        {
            var specialUser = new ThisIsTheEntity();

            TypeDescriptor.AddProvider(
                new AssociatedMetadataTypeTypeDescriptionProvider(
                    typeof(ThisIsTheEntity),
                    typeof(ThisIsTheBuddy)),
                specialUser);

            var results = new List<ValidationResult>();
            Console.WriteLine(Validator.TryValidateObject(specialUser, new ValidationContext(specialUser), results, true));
            Console.WriteLine(results.Count);
        }
    }
    
    public class ThisIsTheEntity
    {
        public string Username { get; set; }
    }

    public class ThisIsTheBuddy
    {
        [Required]
        public string Username { get; set; }
    }
}
```